### PR TITLE
Summary at the top of the address page

### PR DESCRIPTION
**Pull request by McEdward**
Resolving issue #121

### CODE REVIEW

**Comment by raedah** [comment]
This test well for me. We are still using float for the templates here. Will be able to move to atoms in a separate PR.

**Comment by chappjc** [comment]
got a panic:

```2017/10/20 18:51:06 "GET http://127.0.0.1:7777/explorer/address/DsdFRC1pgKLxRVqMQnNGjnLyTWSfidv9VWy HTTP/1.1" from 127.0.0.1:47400 - panic: runtime error: invalid memory address or nil pointer dereference
2017/10/20 18:51:06 goroutine 92 [running]:
runtime/debug.Stack(0xc42800ef00, 0x122d180, 0xc425d30ea0)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0xa7
github.com/dcrdata/dcrdata/vendor/github.com/go-chi/chi/middleware.Recoverer.func1.1(0xc42800ef00, 0x7fbd6867a2e8, 0xc427f19380)
	/home/jon/go/src/github.com/dcrdata/dcrdata/vendor/github.com/go-chi/chi/middleware/recoverer.go:25 +0x9c
panic(0xa9fee0, 0x11f50d0)
	/usr/lib/go/src/runtime/panic.go:491 +0x283
github.com/dcrdata/dcrdata/dcrsqlite.makeExplorerAddressTx(0xc420281f80, 0xc42001dfb1, 0x23, 0xc4264ff000)
	/home/jon/go/src/github.com/dcrdata/dcrdata/dcrsqlite/apisource.go:708 +0x19f
github.com/dcrdata/dcrdata/dcrsqlite.(*wiredDB).GetExplorerAddress(0xc420248060, 0xc42001dfb1, 0x23, 0x1f4, 0xc4253d81e0)
	/home/jon/go/src/github.com/dcrdata/dcrdata/dcrsqlite/apisource.go:925 +0x342
github.com/dcrdata/dcrdata/explorer.(*explorerUI).addressPage(0xc428011900, 0x7fbd6867a2e8, 0xc427f19380, 0xc42800f000)
	/home/jon/go/src/github.com/dcrdata/dcrdata/explorer/explorer.go:143 +0xce
github.com/dcrdata/dcrdata/explorer.(*explorerUI).(github.com/dcrdata/dcrdata/explorer.addressPage)-fm(0x7fbd6867a2e8, 0xc427f19380, 0xc42800f000)
	/home/jon/go/src/github.com/dcrdata/dcrdata/explorer/explorer.go:387 +0x48```

File: dcrsqlite/apisource.go [hunk 1]
```diff
@@ -704,6 +704,16 @@ func makeExplorerAddressTx(data *dcrjson.SearchRawTransactionsResult) *explorer.
 	t := time.Unix(tx.Time, 0)
 	tx.FormattedTime = t.Format("1/_2/06 15:04:05")
 	tx.Confirmations = data.Confirmations
+	for i := 0; i < len(data.Vin); i++ {
+		if data.Vin[i].PrevOut.Addresses[0] == address {
```
Comment (chappjc) [inline dcrsqlite/apisource.go]: panic here

File: dcrsqlite/apisource.go [hunk 2]
```diff
@@ -704,6 +704,16 @@ func makeExplorerAddressTx(data *dcrjson.SearchRawTransactionsResult) *explorer.
...
+			tx.SentTotal += *data.Vin[i].AmountIn
+		}
+	}
+	for i := 0; i < len(data.Vout); i++ {
+		if data.Vout[i].ScriptPubKey.Addresses[0] == address {
```
Comment (chappjc) [inline dcrsqlite/apisource.go]: and here

File: dcrsqlite/apisource.go [hunk 3]
```diff
@@ -912,11 +922,40 @@ func (db *wiredDB) GetExplorerAddress(address string, count int) *explorer.Addre
...
 	for _, tx := range txs {
-		addressTxs = append(addressTxs, makeExplorerAddressTx(tx))
+		addressTxs = append(addressTxs, makeExplorerAddressTx(tx, address))
 	}
 
+	NumberOfTx := len(txs)
+	NoOfUnconfirmed := 0
+	var totalreceived, totalsent dcrutil.Amount
+	if NumberOfTx < explorer.AddressRows {
+		for i := 0; i < len(txs); i++ {
```
Comment (chappjc) [inline dcrsqlite/apisource.go]: `for i := range txs {`

Comment (chappjc) [reply dcrsqlite/apisource.go]: or `for _, tx := range txs {` and use `tx` in the loop instead of `txs[i]`

File: dcrsqlite/apisource.go [hunk 4]
```diff
@@ -912,11 +922,40 @@ func (db *wiredDB) GetExplorerAddress(address string, count int) *explorer.Addre
...
+			}
+		}
+	}
+	unspentfunds := totalreceived - totalsent
 	return &explorer.AddressInfo{
-		Address:      address,
-		Transactions: addressTxs,
+		Address:          address,
+		Transactions:     addressTxs,
+		NoOfTransactions: NumberOfTx,
```
Comment (chappjc) [inline dcrsqlite/apisource.go]: `NumTransactions`

File: views/address.tmpl [hunk 5]
```diff
@@ -12,12 +12,24 @@
 
                 <h4>Address</h4>
                 <p class="mono">{{.Address}}</p>
-
+                <p>
+                    {{if eq .NoOfTransactions .AddressRow}}
+                        Currently only the first {{.AddressRow}} transactions are being shown below.
```
Comment (chappjc) [inline views/address.tmpl]: replace "first {{...}}" with "{{...}} most recent"
Also remove "being"

File: views/address.tmpl [hunk 6]
```diff
@@ -12,12 +12,24 @@
...
+                    {{end}}
+                </p>
                 <h5>Transactions</h5>
                 <table class="table table-mono-cells table-sm striped">
                     <thead>
                         <th>Transactions ID</th>
-                        <th class="text-right">Total DCR</th>
+                        <th class="text-right">Received DCR</th>
+                        <th>Sent DCR</th>
+                        <th>Total DCR</th>
```
Comment (chappjc) [inline views/address.tmpl]: What is "Total DCR" ?

Comment (raedah) [reply views/address.tmpl]: Its the total amount sent in the transaction, not address specific.

File: dcrsqlite/apisource.go [hunk 7]
```diff
@@ -706,14 +706,14 @@ func makeExplorerAddressTx(data *dcrjson.SearchRawTransactionsResult, address st
 	tx.Confirmations = data.Confirmations
 
 	for i := range data.Vin {
-		if data.Vin[i].PrevOut.Addresses != nil {
+		if data.Vin[i].PrevOut.Addresses != nil && len(data.Vin[i].PrevOut.Addresses) != 0 {
```
Comment (chappjc) [inline dcrsqlite/apisource.go]: What I meant by "It is OK to check the length of a nil size" is that you can omit the first check and only verify length >0

Comment (chappjc) [reply dcrsqlite/apisource.go]: https://play.golang.org/p/dhx3OR9C5v

Comment (McEdward) [reply dcrsqlite/apisource.go]: Alright, I'll get on that.

**Comment by chappjc** [comment]
Finally tested.  Some observations:

- For some unconfirmed transactions, the Sent DCR can show bogus numbers, usually `-0.00000002`.  This should be handled the same way that the transaction page does it by showing "N/A".
- "DCR" in the column headings is unnecessary clutter.
- I find these columns very confusing: "Received DCR | Sent DCR | Total DCR".  The first two columns are values specific to the address while Total is for the whole transaction.  I think Total DCR is almost irrelevant here.  It's an address summary table, and the user can click the tx to see more anyway.
- The headings "Received DCR" and "Sent DCR" themselves are confusing.  More clear and accurate headings might be simply "to" and "from".  Maybe just "+" and "-".
- The Confirmations column is taking up a lot of space just for the heading.  Lets shorten it.  Maybe "Confirms".

**Comment by chappjc** [comment]
Panic at http://127.0.0.1:7777/explorer/address/DskLpSvj8Mcn2Qf78pyUArBc6B9nKZEkCNr

```panic: runtime error: invalid memory address or nil pointer dereference
2017/10/23 23:45:44 goroutine 864 [running]:
runtime/debug.Stack(0xc42b18ba00, 0x122f180, 0xc4280ee800)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0xa7
github.com/dcrdata/dcrdata/vendor/github.com/go-chi/chi/middleware.Recoverer.func1.1(0xc42b18ba00, 0x7f52842f7ac8, 0xc427e84dc0)
	/home/jon/go/src/github.com/dcrdata/dcrdata/vendor/github.com/go-chi/chi/middleware/recoverer.go:25 +0x9c
panic(0xaa1440, 0x11f70d0)
	/usr/lib/go/src/runtime/panic.go:491 +0x283
github.com/dcrdata/dcrdata/dcrsqlite.makeExplorerAddressTx(0xc42c18d280, 0xc42af0a2d1, 0x23, 0xc42902e3d0)
	/home/jon/go/src/github.com/dcrdata/dcrdata/dcrsqlite/apisource.go:710 +0x1b1
github.com/dcrdata/dcrdata/dcrsqlite.(*wiredDB).GetExplorerAddress(0xc4201d54a0, 0xc42af0a2d1, 0x23, 0x1f4, 0xc42902e2d0)
	/home/jon/go/src/github.com/dcrdata/dcrdata/dcrsqlite/apisource.go:949 +0x33f
github.com/dcrdata/dcrdata/explorer.(*explorerUI).addressPage(0xc426864780, 0x7f52842f7ac8, 0xc427e84dc0, 0xc42b18bb00)
	/home/jon/go/src/github.com/dcrdata/dcrdata/explorer/explorer.go:144 +0xce
github.com/dcrdata/dcrdata/explorer.(*explorerUI).(github.com/dcrdata/dcrdata/explorer.addressPage)-fm(0x7f52842f7ac8, 0xc427e84dc0, 0xc42b18bb00)
	/home/jon/go/src/github.com/dcrdata/dcrdata/explorer/explorer.go:418 +0x48
```

**Comment by chappjc** [comment]
This needs some styling:

image

Emulate how the other pages show such information.  For example:

image

Time, size, fee, etc above look presentable.

**Comment by raedah** [comment]
Interestingly, when I was analyzing transactions on the blockchain earlier, the Total column came in quite handy. I also would have liked to have seen the input and output addresses. Lets keep this view as the simplified view and remote the Total column, and create a new detailed address view that still has it.

> "DCR" in the column headings is unnecessary clutter.
Are you suggesting we make it the suffix on each of the amounts instead?

>The headings "Received DCR" and "Sent DCR" themselves are confusing.
I think Credit and Debit might be what we are looking for. To make it more obvious we could do +green_amount and -red_amount. It could also be merged into a single column for the simplified view, and keep it as two columns for the detailed view so it can been seen when a txs is sending funds back to the same address.

**Comment by chappjc** [comment]
I guess having DCR in the header is good.  I like Credit/Debit as that's more what we're conveying, and I think we should definitely demo the single column idea with +green and -red.  That could be just right here!  It would be really odd to have a tx that pays to an address and redeems an output with that same address, and if there is, they can click through to see that.

File: dcrsqlite/apisource.go [hunk 8]
```diff
@@ -958,15 +958,19 @@ func (db *wiredDB) GetExplorerAddress(address string, count int) *explorer.Addre
...
+						t, _ := dcrutil.NewAmount(y.Value)
+						totalreceived += t
+					}
 				}
 			}
 			for _, u := range tx.Vin {
-				if address == u.PrevOut.Addresses[0] {
-					t, _ := dcrutil.NewAmount(*u.AmountIn)
-					totalsent += t
+				if len(u.PrevOut.Addresses) != 0 {
```
Comment (chappjc) [inline dcrsqlite/apisource.go]: Note what I said about the panic.  Seems we also need `u.PrevOut!=nil`

Comment (McEdward) [reply dcrsqlite/apisource.go]: In addition to Len(u.PrevOut.Addresses) != 0?

Comment (chappjc) [reply dcrsqlite/apisource.go]: Yeah, turns out that PrevOut being nil was the issue

Comment (McEdward) [reply dcrsqlite/apisource.go]: Alright, would that be all before I make the next commit?

Comment (chappjc) [reply dcrsqlite/apisource.go]: I think so, but we'll keep iterating if needed.  No biggie.

`if u.PrevOut != nil && len(u.PrevOut.Addresses) > 0 {`

**Pull request merged**